src/protobuf-net.Grpc.Reflection/FileDescriptorSetFactory.cs

```diff
@@ -0,0 +1,103 @@
+
+namespace ProtoBuf.Grpc.Reflection
+{
+    internal static class FileDescriptorSetFactory
+    {
+        internal static FileDescriptorSet Create(IEnumerable<Type> serviceTypes, BinderConfiguration? binderConfiguration = null)
+        {
+            var fileDescriptorSet = new FileDescriptorSet();
+            var configuration = binderConfiguration ?? BinderConfiguration.Default;
+
+            foreach (var serviceType in serviceTypes)
+            {
+                Populate(fileDescriptorSet, serviceType, configuration);
+            }
+
+            fileDescriptorSet.Process();
+            return fileDescriptorSet;
+        }
+
+        private static void Populate(FileDescriptorSet fileDescriptorSet, Type serviceType, BinderConfiguration binderConfiguration)
+        {
+            string? serviceName, inputFile, outputFile;
+            var serviceContracts = typeof(IGrpcService).IsAssignableFrom(serviceType)
+                ? new HashSet<Type> { serviceType }
+                : ContractOperation.ExpandInterfaces(serviceType);
+
+            foreach (var serviceContract in serviceContracts)
+            {
+                if (!binderConfiguration.Binder.IsServiceContract(serviceContract, out serviceName)) continue;
+
+                var serviceDescriptor = new ServiceDescriptorProto
+                {
+                    Name = serviceName!.Split('.').Last()
+                };
+
+                var dependencies = new HashSet<string>();
+                foreach (var op in ContractOperation.FindOperations(binderConfiguration, serviceContract))
+                {
+                    // TODO: Validate op
+                    serviceDescriptor.Methods.Add(new MethodDescriptorProto
+                    {
+                        Name = op.Name,
+                        InputType = GetType(op.From, fileDescriptorSet, out inputFile),
+                        OutputType = GetType(op.To, fileDescriptorSet, out outputFile),
+                        ClientStreaming = op.MethodType == MethodType.ClientStreaming ||
+                                          op.MethodType == MethodType.DuplexStreaming,
+                        ServerStreaming = op.MethodType == MethodType.ServerStreaming ||
+                                          op.MethodType == MethodType.DuplexStreaming
+                    });
+
+                    dependencies.Add(inputFile);
+                    dependencies.Add(outputFile);
+                }
+
+                var fileDescriptor = new FileDescriptorProto
+                {
+                    Name = serviceName + ".proto",
+                    Services = { serviceDescriptor },
+                    Syntax = "proto3",
+                    Package = serviceContract.Namespace
+                };
+
+                foreach (var dependency in dependencies)
+                {
+                    fileDescriptor.Dependencies.Add(dependency);
+                    fileDescriptor.AddImport(dependency, true, default);
```

Comment:
I noticed I needed to `AddImport` in order for `FileDescriptorSet.Process()` to resolve types from dependencies. Shouldn't it be enough to add the dependency to `Dependencies`?